src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Indexer/Indexer/SearchQueryBase.cs

```diff
@@ -0,0 +1,382 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Numerics;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.CmdPal.Ext.Indexer.Indexer.OleDB;
+using Microsoft.CmdPal.Ext.Indexer.Indexer.Propsys;
+using Microsoft.CmdPal.Ext.Indexer.Native;
+using Microsoft.CmdPal.Ext.Indexer.Utils;
+using MSDASC;
+
+namespace Microsoft.CmdPal.Ext.Indexer.Indexer;
+
+internal abstract class SearchQueryBase
+{
+    private readonly Lock _lockObject = new(); // Lock object for synchronization
+
+    public uint ReuseWhereID { get; set; }
+
+    public uint NumResults { get; set; }
+
+    private IRowset currentRowset;
+
+    private IRowset reuseRowset;
+
+    public SearchQueryBase()
+    {
+    }
+
+    protected virtual void FetchRows(out ulong totalFetched)
+    {
+        ulong fetched = 0;
+
+        var getRow = currentRowset as IGetRow;
+        uint rowCountReturned;
+
+        ulong batchSize = 5000; // Number of rows to fetch in each batch
+
+        do
+        {
+            IntPtr prghRows;
+            var res = currentRowset.GetNextRows(
+                IntPtr.Zero,       // hReserved
+                0,                 // lRowsOffset
+                (long)batchSize,         // cRows
+                out rowCountReturned,  // pcRowsObtained
+                out prghRows);
+
+            if (res < 0)
+            {
+                Logger.LogError($"Error fetching rows: {res}");
+                break;
+            }
+
+            if (rowCountReturned == 0 || rowCountReturned == IntPtr.Zero)
+            {
+                // No more rows to fetch
+                break;
+            }
+
+            // Marshal the row handles
+            var rowHandles = new IntPtr[rowCountReturned];
+            Marshal.Copy(prghRows, rowHandles, 0, (int)rowCountReturned);
+
+            ULongLongAdd(fetched, rowCountReturned, out fetched);
+
+            for (ulong i = 0; i < rowCountReturned; i++)
+            {
+                object unknown;
+                res = getRow.GetRowFromHROW(null, rowHandles[i], typeof(IPropertyStore).GUID, out unknown);
+                if (res != 0)
+                {
+                    Logger.LogError($"Error getting row from HROW: {res}");
+                    break;
+                }
+
+                var propStore = (IPropertyStore)unknown;
+
+                NumResults++;
+                OnFetchRowCallback(propStore);
+            }
+
+            // Release the rows
+            res = currentRowset.ReleaseRows(
+                rowCountReturned,
+                rowHandles,
+                IntPtr.Zero,
+                null,
+                null);
+
+            if (res != 0)
+            {
+                Logger.LogError($"Error releasing rows: {res}");
+                break;
+            }
+
+            // Free the memory allocated for row handles
+            Marshal.FreeCoTaskMem(prghRows);
```

Comment:
This seems to be the line that's throwing the heap corruption. Yea of course just commenting it out _works_, but leaks memory. I wonder why this is throwing - `prghRows` looked like it had a valid non-zero value so I don't what's up here